### PR TITLE
[SPARK-25771][PYSPARK]Fix improper synchronization in PythonWorkerFactory

### PR DESCRIPTION
**Pull request by zsxwing**
## What changes were proposed in this pull request?

Fix the following issues in PythonWorkerFactory
1. MonitorThread.run uses a wrong lock.
2. `createSimpleWorker` misses `synchronized` when updating `simpleWorkers`.

Other changes are just to improve the code style to make the thread-safe contract clear.

## How was this patch tested?

Jenkins

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala [hunk 1]
```diff
@@ -278,7 +289,7 @@ private[spark] class PythonWorkerFactory(pythonExec: String, envVars: Map[String
 
     override def run() {
       while (true) {
-        synchronized {
+        self.synchronized {
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: this is fix 1

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala [hunk 2]
```diff
@@ -163,7 +172,9 @@ private[spark] class PythonWorkerFactory(pythonExec: String, envVars: Map[String
       try {
         val socket = serverSocket.accept()
         authHelper.authClient(socket)
-        simpleWorkers.put(socket, worker)
+        self.synchronized {
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: this is fix 2.

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala [hunk 3]
```diff
@@ -31,15 +32,15 @@ import org.apache.spark.security.SocketAuthHelper
...
+  extends Logging { self =>
 
   import PythonWorkerFactory._
 
   // Because forking processes from Java is expensive, we prefer to launch a single Python daemon,
   // pyspark/daemon.py (by default) and tell it to fork new workers for our tasks. This daemon
   // currently only works on UNIX-based systems now because it uses signals for child management,
   // so we can also fall back to launching workers, pyspark/worker.py (by default) directly.
-  val useDaemon = {
+  private val useDaemon = {
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: Why are we fixing this? Looks not directly related to the fix and this is already private class.

Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: `daemonModule` and `workerModule` too.

**Comment by HyukjinKwon** [comment]
Please feel `How was this patch tested?` as well in the PR description.

**Comment by HyukjinKwon** [comment]
Please fill `How was this patch tested?` as well in the PR description.

**Review by HyukjinKwon — APPROVED** [review]

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala (hunk 3 above)
Comment (zsxwing) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: Fixing these since I'm touching a lot of fields in this file.

**Comment by zsxwing** [comment]
Thanks for reviewing this. Merging to master.

**Pull request closed without merging**